elastio-s3-changelog/cloudformation-single.yaml

```diff
@@ -0,0 +1,263 @@
+    Type: AWS::SQS::Queue
+    Properties:
+      MessageRetentionPeriod: 1209600 # 14 days
+      VisibilityTimeout: 43200 # 12 hours
+      RedrivePolicy:
+        deadLetterTargetArn: !GetAtt dlq.Arn
+        maxReceiveCount: 3
+      Tags:
+        - Key: elastio:resource
+          Value: 'true'
+        - Key: elastio:s3-changelog-bucket
+          Value: !Ref BucketName
+        - Key: elastio:s3-changelog-stack-arn
+          Value: !Ref AWS::StackId
+
+  dlq:
+    Type: AWS::SQS::Queue
+    Properties:
+      MessageRetentionPeriod: 1209600
+      VisibilityTimeout: 43200
+      Tags:
+        - Key: elastio:resource
+          Value: 'true'
+        - Key: elastio:s3-changelog-dlq
+          Value: !Ref BucketName
+        - Key: elastio:s3-changelog-stack-arn
+          Value: !Ref AWS::StackId
+
+  eventRule:
+    Type: AWS::Events::Rule
+    DependsOn: queue
+    Properties:
+      Description: !Sub Track new objects creation in ${BucketName}
+      EventPattern: !If
+        - NoKeyPrefixes
+        - !Sub >
+          {
+            "source": ["aws.s3"],
+            "detail-type": ["Object Created"],
+            "detail": { "bucket": { "name": ["${BucketName}"] } }
+          }
+        - !Join
+          - ''
+          - - !Sub >
+              {
+                "source": ["aws.s3"],
+                "detail-type": ["Object Created"],
+                "detail":
+                {
+                  "bucket": { "name": ["${BucketName}"] },
+                  "object": { "key": [
+            - !Sub
+              - '{ "prefix": "${prefixes}" }'
+              - prefixes: !Join
+                - '" },{ "prefix": "'
+                - !Split [',', !Ref KeyPrefixes]
+            - '] } } }'
+      Targets:
+        - Arn: !GetAtt queue.Arn
+          Id: changelog-queue
+          InputTransformer:
+            InputPathsMap:
+              key: '$.detail.object.key'
+              type: '$.detail-type'
+            InputTemplate:
+              '{
+                "objects": [{
+                  "key": <key>,
+                  "event_type": <type>
+                }]
+              }'
+
+  queueEventBridgePolicy:
+    Type: AWS::SQS::QueueInlinePolicy
+    Properties:
+      Queue: !Ref queue
+      PolicyDocument:
+        Statement:
+          - Effect: Allow
+            Action:
+              - sqs:SendMessage
+            Resource: !GetAtt queue.Arn
+            Principal:
+              Service: events.amazonaws.com
+            Condition:
+              ArnEquals:
+                'aws:SourceArn': !GetAtt eventRule.Arn
+
+  ssmParameter:
+    Type: AWS::SSM::Parameter
+    Properties:
+      Name: !Sub /elastio/s3/changelog/${BucketName}/${queue.QueueName}
+      Type: String
+      Value: !Sub
+        - >
+          {
+            "prefixes": [${prefixes}],
+            "qrts": ${EnableQrts}
+          }
+        - prefixes: !If
+          - NoKeyPrefixes
+          - ''
+          - !Sub
+            - '"${list}"'
```

Comment:
Will this generate `"foo/,bar/,baz/"`? Let's make it into a JSON array instead

Reply:
no, this will generate `["foo/","bar/","baz/"]`

Reply:
I still don't understand this logic, and it looks wrong to me. I wrote you in PM

Reply:
I didn't notice the double quotes around comma separator here. Now it's clear, thanks